### ISSUE TITLE
Apple MPS support feedback

### ISSUE THREAD

**Issue by leejr72** [suno-ai/bark#266]
Seen the older issues on the topic issue#22, Understanding the MPS is experimental are the results below expected performance?  Please let me know if not implementing correctly.

Happy to refactor and rerun tests in supporting dev of MPS feature.

![bark_MPS_True_1](https://user-images.githubusercontent.com/110568703/236720947-07d674de-70b3-4648-b48a-a81f84a678bf.png)
![bark_MPS_True_2](https://user-images.githubusercontent.com/110568703/236720960-26218216-4186-4cb9-a9c5-a2938cb3b25d.png)
![bark_MPS_False_1](https://user-images.githubusercontent.com/110568703/236720973-6252e5f7-3754-4a07-a604-c27d0ac6661e.png)
![bark_MPS_False_2](https://user-images.githubusercontent.com/110568703/236720978-90c15250-fb7a-4106-a028-a26208231bbc.png)

**Comment by gkucsko** [comment]
hmm, i think when i tried i saw something around 1.5-2x speedup vs CPU on an M1 Max

**Comment by leejr72** [comment]
Any special process to enabling.  Besides adding the os import mps = true?  Will try and find a tool to measure mps usage and run some longer samples.  Thank you for confirming should be seeing increased performance.

**Comment by gkucsko** [comment]
much appreciated!

**Comment by leejr72** [comment]
All testing was done with the default Jupiter long form generator example.  Except added the MPS = True/False
Is it supposed to turn off MPS when set to false?
Dose the very first time executing take longer?

love the project and wish to help in any manor.  Let me know if can help test further or run different testing for the project.

<img width="660" alt="First run with MPS True" src="https://github.com/suno-ai/bark/assets/110568703/9ef3f1c5-bb6c-4cd8-8950-2767d67b1790">
<img width="683" alt="Second run with MPS False" src="https://github.com/suno-ai/bark/assets/110568703/12b3b0d2-8d4f-4596-ae92-455679fbe491">
<img width="1496" alt="Third run with MPS True" src="https://github.com/suno-ai/bark/assets/110568703/6f377be8-b78d-4217-bd2c-0b5099509139">
<img width="672" alt="Fourth run with MPS False" src="https://github.com/suno-ai/bark/assets/110568703/fbaf2a45-de00-4bd0-bfc6-54615b084b5b">
<img width="656" alt="CPU MPS True and False" src="https://github.com/suno-ai/bark/assets/110568703/3c49d275-8ea5-4f20-a96f-0379593a485d">
<img width="346" alt="GPU MPS True and False" src="https://github.com/suno-ai/bark/assets/110568703/98f50215-2883-48a8-91e1-1ebb311ad62f">

**Comment by gkucsko** [comment]
so you are seeing the same speed? did you restart the kernel inbetween and set the env var before doing anything else?

**Comment by gkucsko** [comment]
will try to run on my m1 later when i have better internet, but when i checked last i saw
- like you mentioned some weird warmup seems to be needed where on M1 a first inference run is slow
- ~2x faster on M1 vs cpu

**Comment by gkucsko** [comment]
<img width="804" alt="image" src="https://github.com/suno-ai/bark/assets/5068315/03e89466-d6ff-4a87-9da0-0425b70bc405">

**Comment by gkucsko** [comment]
yeah so after some warmup i'm seeing like 2x faster on MPS (this is M1 Max)

**Comment by leejr72** [comment]
Trying to repro however something happened to were my code no longer uses the MPS.
Reinstalled following with the following commands
conda create -n bark python=3.10
conda activate bark
git clone https://github.com/suno-ai/bark
cd bark
pip install . 
pip uninstall -y torch torchvision torchaudio 
pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cpu

`from bark import SAMPLE_RATE, generate_audio, preload_models
from scipy.io.wavfile import write as write_wav
from IPython.display import Audio

import os
os.environ["SUNO_ENABLE_MPS"] = "True"
torch.device("mps")

# download and load all models
preload_models()

# check environment
print(f"Python Platform: {platform.platform()}")
print(f"Python {sys.version}")
print(f"PyTorch Version: {torch.__version__}")
print("GPU is", "available" if has_gpu else "NOT AVAILABLE")
print("MPS (Apple Metal) is", "AVAILABLE" if has_mps else "NOT AVAILABLE")
print(f"Target device is {device}")
print(f"Torch Device: {torch.device()}")
print(os.environ["SUNO_ENABLE_MPS"])

# generate audio from text
text_prompt = """
     Hello, my name is Suno. And, uh — and I like pizza. [laughs] 
     But I also have other interests such as playing tic tac toe.
"""
audio_array = generate_audio(text_prompt)

# save audio to disk
write_wav("bark_generation.wav", SAMPLE_RATE, audio_array)
  
# play text in notebook
Audio(audio_array, rate=SAMPLE_RATE)`

Am I missing a required command?
![Screenshot 2023-05-12 at 1 49 39 PM](https://github.com/suno-ai/bark/assets/110568703/61d6593d-6c3a-4820-a8b6-2e2c0e8bd671)
<img width="641" alt="Screenshot 2023-05-12 at 2 28 23 PM" src="https://github.com/suno-ai/bark/assets/110568703/79795851-403a-4428-9fc2-1ce40bfd0523">
![Screenshot 2023-05-12 at 5 21 27 PM](https://github.com/suno-ai/bark/assets/110568703/7b521fce-cb17-4f2b-8e19-b873b8159e28)
<img width="346" alt="Screenshot 2023-05-12 at 5 22 22 PM" src="https://github.com/suno-ai/bark/assets/110568703/c553cfe6-eaf1-46a3-9125-bf072181aa6c">

**Comment by gkucsko** [comment]
you have to set the environment variable before anything else gets imported, could that be the reason?

**Comment by leejr72** [comment]
Thank you for your help new to the python.  Got everything running and results below.  Seems we are seeing similar performance.  First time MPS little slow however pickups.  Over all seeing 2x improvements.  Thank you please let me know if I can help with any other testing.

Quick question when executing the "pip install ." how do I install the dev section also?

Also received the following when I tried to use conda to install the nightly version of PyTorch 
  File "/opt/homebrew/Caskroom/miniconda/base/envs/bark-infinity/lib/python3.10/site-packages/torch/nn/functional.py", line 2210, in embedding
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
IndexError: index out of range in self

wiped out the conda environment and rebuild following your instructions using pip

Install method
git clone https://github.com/suno-ai/bark
cd bark
conda create -n bark python=3.10
conda activate bark
pip install . 
pip uninstall -y torch torchvision torchaudio 
pip install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cpu
I also have to pip install nltk to split up the sentences.

Python code used for testing.
import os
os.environ["SUNO_ENABLE_MPS"] = "True"
os.environ["SUNO_OFFLOAD_CPU"] = "False"
os.environ["SUNO_USE_SMALL_MODELS"] = "False"

from bark import SAMPLE_RATE, generate_audio, preload_models
from scipy.io.wavfile import write as write_wav
from IPython.display import Audio
from datetime import datetime as dt
import numpy as np
import nltk  # we'll use this to split into sentences
nltk.download('punkt')

preload_models()

text_prompt = """
     Hello, I am called BARK and am a new text to audio model made by SUNO!
     Let me read an excerpt from War of the Worlds to you. [clears throat]
"""

audio_array = generate_audio(text_prompt, history_prompt="v2/en_speaker_6")

![Screenshot 2023-05-13 at 7 31 33 PM](https://github.com/suno-ai/bark/assets/110568703/f4af1a43-ab3c-4596-a1c5-d6558b63dfbe)

<img width="346" alt="Screenshot 2023-05-13 at 7 05 22 PM" src="https://github.com/suno-ai/bark/assets/110568703/8f8787a0-064d-4f45-b58a-5096ce9ac334">

<img width="770" alt="Screenshot 2023-05-13 at 7 27 17 PM" src="https://github.com/suno-ai/bark/assets/110568703/3d0b4b9b-e082-48da-936b-9f3ba4c07c79">

<img width="1486" alt="Screenshot 2023-05-13 at 7 27 35 PM" src="https://github.com/suno-ai/bark/assets/110568703/e7480ff6-a30a-4217-91f7-0ac5e8e4209e">

**Comment by keldenl** [comment]
On M1 pro 32gb ram, Bark works on CPU. 

When trying MPS, I'm getting the self index out of bounds error when not using PyTorch/other dependencies nightly, but when I install nightly it just hangs for at least 5 minutes.

I'm on python3.11 – any tips on getting MPS working?

**Comment by gkucsko** [comment]
this should be a super simple error to fix (1-liner), but i can't do it since i can't reproduce it :/ would love help here from anyone who can

**Comment by japanvik** [comment]
@keldenl It was also hanging for me, but it worked for me if I set
SUNO_OFFLOAD_CPU = False

**Comment by fiq** [comment]
Not tried nightly, can have a look when I have time. I've been using bark for a while on an M2Pro with the following. There are CPU fallback scenarios, but that hasn't prevented generation in reasonable time:
```
SUNO_ENABLE_MPS=True
PYTORCH_ENABLE_MPS_FALLBACK=1
```
